### DISCUSSION TITLE
Explicit Hs in a SMILE do not get an atom Id

### DISCUSSION THREAD

**Discussion by chymotrypsin13** [rdkit/rdkit#4585] [Q&A]
Hi All,

I am working on a project involving drawing different pieces of a molecule based on slicing the SMILE string (not necessarily generating fragments or SMARTS).

I have found a molecule with lots of stereocenters where the Hs at these stereocenters are included in the SMILEs.

smiles_can = df_trim['CAN SMILES'][1]
smiles_non = df_trim['SMILES'][1]

![image](https://user-images.githubusercontent.com/92038765/136221570-c4ae3459-baac-4888-b5bd-9fb26c7c1212.png)

comparing the SMILES from the molecule ID (smiles_non) vs the canonicalized SMILES (smiles_can). From the SMILES string there are 8 Hs explicitly defined.

If I draw the molecule with:
m = Chem.MolFromSmiles(smiles_can, sanitize = False)

![image](https://user-images.githubusercontent.com/92038765/136222078-250af421-64da-4cb6-bc64-a49eda0f1cc9.png)

I get this molecule with 53 numbered atoms and 8 explicit Hs.

Now I try to get atom order from the Molecule so I can assign atom # to each atom in the SMILE string:
order = eval(m.GetProp('_smilesAtomOutputOrder'))
order goes from 0 to 48 and len(order) = 49.

Next is to get atom Id with:
for atom in m.GetAtoms():
    print(atom.GetIdx(), atom.GetSymbol())

which yields 3 Hs with Id 0, 16, and 23; meaning H # 49, 50, 51, 52, and 53 are missing
If I then convert this molecule back into a SMILE:
t = Chem.MolToSmiles(m)
![image](https://user-images.githubusercontent.com/92038765/136223108-dd533f8d-c753-45db-98cc-344368653485.png)
the output SMILE is exactly the same as smiles_can that was used to make the molecule.

Next I tried sanitizing the molecule:
![image](https://user-images.githubusercontent.com/92038765/136224631-98398600-892f-4fd8-87c7-b2eecd1108c0.png)

which shows 4 explicit Hs (# 46, 47, 48, 49)

![image](https://user-images.githubusercontent.com/92038765/136224110-15ab0cdc-168d-46e3-969f-70e5a01d911d.png)
but the output smile string again has 8 explicit Hs. 
using order_new = eval(m2.GetProp('_smilesAtomOutputOrder'))
order_new goes from 0 to 45 and len(order_new) = 46.

and again with:
for atom in m2.GetAtoms():
    print(atom.GetIdx(), atom.GetSymbol())
yielding 0 H's in the atom ID list

What is the reason for the missing hydrogens in the atom count and atom Id?

My initial thought was simply to take all the input SMILES and add explicit hydrogens, but for large datasets that would be computationally more expensive for the modeling piece of the project.

Any help would be much appreciated, I have been struggling with this for days.
rdkit version 2020.03.2

**Comment by greglandrum** [comment]
MolFromSmiles removes Hs in the input SMILES. If you want to preserve them.
Here’s a link to some code which shows how this works:
https://github.com/rdkit/rdkit/blob/d20e5cadc81bf6c7b4e590124866f178f2f2fe28/Code/GraphMol/Wrap/rough_test.py#L5965
